File: POSmaisgestao/app/src/test/java/com/pos_mais_gestao/hardware/balanca/protocol/Prt3ParserTest.java
package com.pos_mais_gestao.hardware.balanca.protocol;

import static org.junit.Assert.assertArrayEquals;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import com.pos_mais_gestao.domain.balanca.BalancaEstado;
import com.pos_mais_gestao.domain.balanca.BalancaLeitura;
import java.nio.charset.StandardCharsets;
import java.util.List;
import org.junit.Test;

public class Prt3ParserTest {
    @Test
    public void lePesoNormalEZero() {
        BalancaLeitura normal = parse("\u000214385\u0003").get(0);
        BalancaLeitura zero = parse("\u000200000\u0003").get(0);

        assertEquals("14.385", normal.getPesoKg().toPlainString());
        assertTrue(normal.isPesoValido());
        assertEquals("0.000", zero.getPesoKg().toPlainString());
        assertFalse(zero.isPesoValido());
    }

    @Test
    public void distingueEstadosDaBalanca() {
        assertEquals(BalancaEstado.INSTAVEL, parse("\u0002IIIII\u0003").get(0).getEstado());
        assertEquals(BalancaEstado.PESO_NEGATIVO, parse("\u0002NNNNN\u0003").get(0).getEstado());
        assertEquals(BalancaEstado.SOBRECARGA, parse("\u0002SSSSS\u0003").get(0).getEstado());
    }

    @Test
    public void rejeitaFrameIncompletoECaracteresInvalidos() {
        assertTrue(parse("\u0002123").isEmpty());
        assertEquals(BalancaEstado.ERRO, parse("\u0002123\u0003").get(0).getEstado());
        assertEquals(BalancaEstado.ERRO, parse("\u000212A45\u0003").get(0).getEstado());
    }

    @Test
    public void processaMultiplosFramesERuido() {
        List<BalancaLeitura> leituras = parse("lixo\u000200100\u0003xx\u000200250\u0003");

        assertEquals(2, leituras.size());
        assertEquals("0.100", leituras.get(0).getPesoKg().toPlainString());
        assertEquals("0.250", leituras.get(1).getPesoKg().toPlainString());
    }

    @Test
    public void recuperaSincronizacaoEmNovoStxEChunks() {
        Prt3Parser parser = new Prt3Parser();
        byte[] primeiro = "\u0002ruido\u000201".getBytes(StandardCharsets.ISO_8859_1);
        byte[] segundo = "250\u0003".getBytes(StandardCharsets.ISO_8859_1);

        assertTrue(parser.aceitar(primeiro, primeiro.length).isEmpty());
        List<BalancaLeitura> leituras = parser.aceitar(segundo, segundo.length);
        assertEquals(1, leituras.size());
        assertEquals("1.250", leituras.get(0).getPesoKg().toPlainString());
    }

    @Test
    public void comandoDePesoEhEnq() {
        assertArrayEquals(new byte[] {0x05}, Prt3Protocol.solicitarPeso());
    }

    private List<BalancaLeitura> parse(String valor) {
        byte[] bytes = valor.getBytes(StandardCharsets.ISO_8859_1);
        return new Prt3Parser().aceitar(bytes, bytes.length);
    }
}

File: estacao-balanca/test/prt3_parser_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:estacao_balanca/features/balanca/balanca_leitura.dart';
import 'package:estacao_balanca/features/balanca/prt3_parser.dart';
import 'package:estacao_balanca/features/balanca/prt3_protocol.dart';

List<BalancaLeitura> parse(String valor) {
  final bytes = latin1.encode(valor);
  return Prt3Parser().aceitar(bytes);
}

void main() {
  test('le peso normal e zero', () {
    final normal = parse('\u000214385\u0003').first;
    final zero = parse('\u000200000\u0003').first;

    expect(normal.pesoKg!.toStringAsFixed(3), '14.385');
    expect(normal.isPesoValido, isTrue);
    expect(zero.pesoKg!.toStringAsFixed(3), '0.000');
    expect(zero.isPesoValido, isFalse);
  });

  test('distingue estados da balanca', () {
    expect(parse('\u0002IIIII\u0003').first.estado, BalancaEstado.instavel);
    expect(
      parse('\u0002NNNNN\u0003').first.estado,
      BalancaEstado.pesoNegativo,
    );
    expect(parse('\u0002SSSSS\u0003').first.estado, BalancaEstado.sobrecarga);
  });

  test('rejeita frame incompleto e caracteres invalidos', () {
    expect(parse('\u0002123'), isEmpty);
    expect(parse('\u0002123\u0003').first.estado, BalancaEstado.erro);
    expect(parse('\u000212A45\u0003').first.estado, BalancaEstado.erro);
  });

  test('processa multiplos frames e ruido', () {
    final leituras = parse('lixo\u000200100\u0003xx\u000200250\u0003');
    expect(leituras.length, 2);
    expect(leituras[0].pesoKg!.toStringAsFixed(3), '0.100');
    expect(leituras[1].pesoKg!.toStringAsFixed(3), '0.250');
  });

  test('recupera sincronizacao em novo STX e chunks', () {
    final parser = Prt3Parser();
    final primeiro = latin1.encode('\u0002ruido\u000201');
    final segundo = latin1.encode('250\u0003');

    expect(parser.aceitar(primeiro), isEmpty);
    final leituras = parser.aceitar(segundo);
    expect(leituras.length, 1);
    expect(leituras.first.pesoKg!.toStringAsFixed(3), '1.250');
  });

  test('comando de peso eh ENQ', () {
    expect(Prt3Protocol.solicitarPeso(), [0x05]);
  });
}
